### ISSUE TITLE
Find operation does not begin searching at cursor

### ISSUE THREAD

**Issue by Gitoffthelawn** [notepad-plus-plus/notepad-plus-plus#7824]
### Description of the Issue
The find operation completely skips all characters in the current string (where the cursor is located) until the first space is encountered.

Originally reported here: https://community.notepad-plus-plus.org/topic/15808/how-can-i-prevent-npp-to-add-any-text-by-itself-in-the-search-string

### Steps to Reproduce the Issue

1. Open NPP and type `verylongstringone verylongstringtwo`
2. Place the cursor just after the first `n`
3. Use the Find dialog box to search for `string`

### Expected Behavior
Find the first occurrence of the string `string`

### Actual Behavior
Misses the first occurrence of the string `string`

### Debug Information

Notepad++ v7.8.2   (32-bit)
Build time : Dec  4 2019 - 01:36:59
Admin mode : OFF
Local Conf mode : ON
OS Name : Windows 7  (64-bit) 
OS Build : 7601.0

**Comment by xylographe** [comment]
Cannot reproduce. (Tested with 32-bit NPP on 64-bit Win7)
Did you, by any chance, have the _Backward direction_ option enabled?

**Comment by Gitoffthelawn** [comment]
Just tested it again.
The *Backward direction* option is definitely disabled.

**Comment by sasumner** [comment]
I'm guessing that step 2a was missing from the reproduction process:

2a.  Press ctrl+f

When that is done, `verylongstringone` will get selected in the editing tab (and placed in the Find window's `Find what` field) and the actual caret position will be after the final `e`.  Thus, a search for `string` will hit in `verylongstringtwo` and NOT in `verylongstringone`.

This issue can be "solved" by using Notepad++ 7.8.3 and later's option to not autopopulate the Find window's `Find what` box when invoking a search with ctrl+f.

**Comment by xylographe** [comment]
@sasumner Thank you, spot on!

**Comment by Gitoffthelawn** [comment]
@sasumner Yes, thank you. I've edited the original issue report to reflect that information.

I think even with the great new option to not autopopulate the Find window, this should get fixed because, when that option is not enabled, it happens every time the the user replaces the autopopulated text.

**Comment by sasumner** [comment]
@Gitoffthelawn What are your ideas for "fixing"?  Please be specific and in the context of how this all currently works.

**Comment by Gitoffthelawn** [comment]
@sasumner If I understand things correctly, when autopopulation is disabled using the new option, this should no longer be an issue.

But when autopopulation is enabled, the caret should remain wherever the user clicked, or at the least, at the beginning of the string that is autoselected.

If this isn't possible due to implementation limitations, perhaps just begin the search operation at the caret position that the user selected before pressing Ctrl+F.

**Comment by sasumner** [comment]
>when autopopulation is disabled using the new option, this should no longer be an issue.

I believe you are correct (but everyone keep reading...for the autopop enabled case)

>when autopopulation is enabled, the caret should remain wherever the user clicked

It can't really do this, as a selection is created for the word used for the autopop, and the caret has to be at one end of the selection or the other (can't be in the middle).  I'm sure the vast majority of users like this selection being made in the doc text.

>or...(the caret should be placed)...at the beginning of the string that is autoselected

This could be done, I suppose; would it bother existing users to have this happen?

>perhaps just begin the search operation at the caret position that the user selected before pressing Ctrl+F

Doable, I suppose.

For the record, I was bothered by the existing behavior myself, almost 2 years ago I noted it in this Community thread:  https://community.notepad-plus-plus.org/topic/15808/how-can-i-prevent-npp-to-add-any-text-by-itself-in-the-search-string

What is the public opinion out there, about changing this?

**Comment by Ekopalypse** [comment]
>It can't really do this, as a selection is created for the word used for the autopop, and the caret has to be at one end of the selection or the other (can't be in the middle)

Are you sure about this? I've to be honest, I didn't check the source code but my first thought was
using SCI_GETCURRENTPOS, SCI_WORDSTARTPOSITION and SCI_WORDENDPOSITION.
So I don't see really a need to have a selection.

If I understand correctly, I would prefer having no selection at all and the cursor stays where I left it when opening the find dialog. Occasionally it happens that I close the find window and start typing
just to find out that I've overwritten the selected word. Not a big deal, nothing ctrl+z couldn't repair
but would be nice if I haven't to use undo operations.
The "disable autopop" :-) of the current word is not really interesting to me as I've trained myself to overwrite it.

**Comment by sasumner** [comment]
>I don't see really a need to have a selection.

There is no actual *need*, but it is one of those things that perhaps legions of current Notepad++ users are used to have happen, and would miss it if it did not do this (and No, let's not have a Preference controlling it!).  

Most other editors that grab the word at the caret and stuff it in the Find box also highlight the word in the main editing window, so I suspect that the user community is really used to such a thing.  Visual Studio does it; okay, not in exactly the same way -- but Notepad++ maybe has more limited options for the *way* to reasonably do it.

**Comment by Gitoffthelawn** [comment]
@donho I just upgraded to Notepad++ v7.8.3.  Thank you _**so very much!!!**_  The new optional Find/Search functionality is such an improvement for me.  Thank you!

**Comment by alankilborn** [comment]
@Gitoffthelawn As the originator of this issue, and as someone that now seems happy (posting just above), can this issue be closed?

**Comment by Gitoffthelawn** [comment]
@alankilborn Thanks for the ping. Since it's been a couple years, I'll add reading everything above to my list and verify that things are resolved.  My quick guess is "yes", but after I take a closer look, I'll be able to provide a definite response.  It might take me a week or two, as I have a few hundred notifications.

**Comment by Gitoffthelawn** [comment]
@alankilborn The issue is resolved if auto-populate for the find dialog box's "Find what" field is disabled.  I re-read all the posts in this thread, and it sounds like the issue does still exist when auto-populate is enabled.

**Comment by alankilborn** [comment]
@Gitoffthelawn said:

>when autopopulation is enabled, the caret should remain wherever the user clicked, 

(With a code change,) the caret _could_ remain where it was; the problem is the user won't be able to see it; this is what the user would see:

![image](https://user-images.githubusercontent.com/77065706/207363082-04f9f143-5def-4d24-a58c-6dafb372ff9c.png)

Now the user is "lost" as to where his search is starting from (imagine a lot of similar text in the editing window).  There is no way to show a caret while the _Find_ window has input focus.

>or at the least, at the beginning of the string that is autoselected

This would break legions of users' workflows, as the first match using the autopopulated value would find the text that the user used to autopopulate (and he already knows about that one and is now annoyed that program isn't going to the NEXT match without another search).  Maybe that value could be "skipped" in the search, but when considering _Replace_, user probably wants the value replaced, not skipped.
This sounds complicated, and risky.

---------------------------

Hmm, some of this ground was already covered by @sasumner back here: https://github.com/notepad-plus-plus/notepad-plus-plus/issues/7824#issuecomment-572767556
